diff --git a/arch/arm/dts/px30-evb.dts b/arch/arm/dts/px30-evb.dts
old mode 100644
new mode 100755
index 089a7ade4e..111a0de0ac
--- a/arch/arm/dts/px30-evb.dts
+++ b/arch/arm/dts/px30-evb.dts
@@ -15,7 +15,7 @@
 	compatible = "rockchip,px30-evb", "rockchip,px30";
 
 	chosen {
-		stdout-path = "serial5:115200n8";
+		stdout-path = "serial2:1500000n8";
 	};
 
 	adc-keys {
@@ -494,7 +494,7 @@
 	status = "okay";
 };
 
-&uart5 {
+&uart2 {
 	status = "okay";
 };
 
diff --git a/arch/arm/dts/px30.dtsi b/arch/arm/dts/px30.dtsi
old mode 100644
new mode 100755
index 0d2325a77f..a136388229
--- a/arch/arm/dts/px30.dtsi
+++ b/arch/arm/dts/px30.dtsi
@@ -398,7 +398,7 @@
 		reg-shift = <2>;
 		reg-io-width = <4>;
 		pinctrl-names = "default";
-		pinctrl-0 = <&uart2m0_xfer>;
+		pinctrl-0 = <&uart2m1_xfer>;
 		status = "disabled";
 	};
 
diff --git a/configs/evb-px30_defconfig b/configs/evb-px30_defconfig
old mode 100644
new mode 100755
index 4a35f3f887..1b82d6aaad
--- a/configs/evb-px30_defconfig
+++ b/configs/evb-px30_defconfig
@@ -10,7 +10,9 @@ CONFIG_TPL_LIBGENERIC_SUPPORT=y
 CONFIG_SPL_DRIVERS_MISC_SUPPORT=y
 CONFIG_NR_DRAM_BANKS=1
 CONFIG_SPL_STACK_R_ADDR=0x600000
-CONFIG_DEBUG_UART_BASE=0xFF178000
+CONFIG_DEBUG_UART_BASE=0xFF160000
+CONFIG_DEBUG_UART_CHANNEL=1
+CONFIG_DEBUG_UART2_CHANNEL=1
 CONFIG_DEBUG_UART_CLOCK=24000000
 CONFIG_DEBUG_UART=y
 CONFIG_TPL_SYS_MALLOC_F_LEN=0x600
